### ISSUE TITLE
Idea: Add a zsh-completions output extension

### ISSUE THREAD

**Issue by jona7o** [bkirwi/decline#250]
I think most of the mac users are using zsh instead of bash. zsh has this awesome plugin for auto completion. I think it could be a great extension for "decline" to autogenerate this zsh-autocompletion output from the Opts. I have seen this output in several CLIs like kubectl for kubernetes (`source <(kubectl completion zsh)`). 

So my suggestion is to build an plugin for decline that generates the needed output for example with `./my-decline-cli completion zsh`. Perhaps as "com.monovore" %% "decline-zsh-completions" % "1.3.0"?

**Comment by bkirwi** [comment]
Yeah, I think this would be a cool enhancement! Some quick thoughts:
- You'd probably want to support bash too at some point. My understanding is that they're not so different, but I haven't looked into it much!
- You could do a pretty reasonable version using just the information we already track... completing option names, etc. If you wanted to complete argument types like paths / enums / etc. we'd need to extend the AST and argument interface. I'm usually very hesitant to make core changes like this, but I'd consider it in this case.
- It should be possible to expose this as a `Command[_] => String` function in a library somewhere. You could then use that to provide the interface you're suggesting... or folks using native packager could hook it into their build and generate completion files automatically maybe.

I'm unlikely to work on this myself, but I'd certainly be open to contributions of this shape! Labelling as such.

**Comment by ziggystar** [comment]
To add to bkirwi: When this will be done, I suggest to make it a general extension with support for multiple shells from the start (even if only one gets implemented initially).

**Comment by morgen-peschke** [comment]
One change that could make it easier to experiment with this and other enhancements may be providing access to some of the `Opt` ADT.
